src/Processors/Transforms/JoiningTransform.h

```diff
@@ -147,18 +152,32 @@ class DelayedJoinedBlocksTransform : public IProcessor
 class DelayedJoinedBlocksWorkerTransform : public IProcessor
 {
 public:
-    explicit DelayedJoinedBlocksWorkerTransform(Block output_header);
+    explicit DelayedJoinedBlocksWorkerTransform(
+        Block left_header_,
+        Block output_header_,
+        size_t max_block_size_,
+        JoinPtr join_);
 
     String getName() const override { return "DelayedJoinedBlocksWorkerTransform"; }
 
     Status prepare() override;
     void work() override;
 
 private:
+    Block left_header;
+    Block output_header;
+    size_t max_block_size;
+    JoinPtr join;
     DelayedBlocksTaskPtr task;
     Chunk output_chunk;
 
-    bool finished = false;
+    /// All joined and non-joined rows from left stream are emitted, only right non-joined rows are left
+    bool left_delayed_stream_finished = false;
```

Comment:
It seems possible to eliminate `left_delayed_stream_finished` flag, - we can leverage `DelayedBlocksTask::finished` flag (so, reseting the task pointer will become unnecessary)

Reply:
`DelayedBlocksTask::finished` means no more buckets, but not the end of current bucket.

Reply:
Yes, I understand. But the flag looks unnecessary.
With this PR,`DelayedJoinedBlocksWorkerTransform` has two stages:
(1) handling delayed blocks for the left table via `task`
(2) handling non-joined blocks from right table for right/full join

So, the `task` is relevant only for stage (1). It's finished on state (2).
Therefore, `task->finished` can be used to understand if stage (1) is completed, -  `IBlockStream::next()` will set `finished` to `true` if there is no more blocks to read, see https://github.com/ClickHouse/ClickHouse/blob/b542fe364ae877aafb365a7b5dddfdefa444b4f5/src/Interpreters/IJoin.h#L114

Reply:
Sorry, I've provided the wrong link to the code, and looking at it confused myself :/ I meant the `DelayedBlocksTask::finished` flag - https://github.com/ClickHouse/ClickHouse/blob/f8f9f86ec6d73eb7a49260c56cca36fa06cae63a/src/Processors/Transforms/JoiningTransform.h#L123